### DISCUSSION TITLE
Change fetchContent behaviour

### DISCUSSION THREAD

**Discussion by rebelchris** [withastro/roadmap#5] [Proposal]
- Start Date: 2021-11-24
- Reference Issues: -
- Implementation PR: -

# Summary

Astro's `fetchContent` currently renders out the HTML for each fetched item.
Causing child components that also use `fetchContent` to break on an infinite loop.

# Example

The current API implementation of the fetchContent always returns rendered HTML.
This means when fetchContent is used inside rendered articles, it causes a infinite loop and a 500 timeout.

You can see this behaviour in the following StackBlitz.
It's causes by the Related component

https://stackblitz.com/edit/github-dchmz1-pdxhaw?file=src%2Fcomponents%2FRelated.astro&on=stackblitz

# Motivation

We should consider changing this behaviour so people can use `fetchContent` inside fetched pages.

This is very common behaviour in blogs for instance.

Sample use-case:

Bob has a blog, each article is rendered with `fetchContent`.
Bob wants to show related articles at the bottom of each article based on tags.
Bob can use `fetchContent` in a child component so he can show 2 related articles.

To achieve this we could refactor the API to NOT render the content by default, or give people the option to turn this on/off.

We could then use a `content.render()` in the page so it will actually render the content slot.

# Detailed design

I need some help here from the core team.

# Drawbacks

- Users have to manually render

# Pros

- Speed would improve
- Gives us the option to load content like related content for example which is quite common in blog articles.

# Alternatives

Alternatively we could leverage a second version of fetchContent? Like fetchRawContent? 
This way existing implementations could keep working.

# Adoption strategy

Depends on the way we go about it.

# Unresolved questions

TBD

**Comment by natemoo-re** [comment]
This is a great suggestion and would solve a lot of problems! FWIW Vite's [`import.meta.glob`](https://vitejs.dev/guide/features.html#glob-import) feature returns functions _to lazily `import()`_ modules for similar reasons.

In 0.21 there's a bit of an API change as well, where `fetchContent` additionally returns a `<Content />` component that will be rendered out. This is to support components-in-markdown and feels conceptually similar to this RFC! We still need to add docs on that feature, but it does use the same eager evaluation approach that you've highlighted as problematic here.

**Reply by rebelchris** [reply]
Nice! 

Do let me know if you would have any other suggestions in how we could achieve the goal of showing fetched content on content pages.
If there are any other workarounds also happy to test those out.

**Comment by rebelchris** [comment]
@natemoo-re @FredKSchott Just tried it on the `0.21.4` version and was surprised to find out it works 🤯

You can see it in action in the stackblitz:
https://stackblitz.com/edit/github-dchmz1-pdxhaw

**Reply by FredKSchott** [reply]
Nice!!!

**Reply by songololo** [reply]
I'm still encountering the circular import loop in v0.23.
